### PR TITLE
support to compare complex numbers #837

### PR DESCRIPTION
**Pull request by gulfaraz**
Allows comparison of Complex Numbers with number, BigNumber, and Complex Number

This behavior is reflected in the following functions
`compare`, `larger`, `largerEq`, `smaller`, `smallerEq`, `max`, `min` and `median`

Note: `median` gives a Complex Number as result if the first argument is a Complex Number

### CODE REVIEW

File: examples/complex_numbers.js [hunk 1]
```diff
@@ -44,6 +44,36 @@ print(c);                      // 1 + i
 var d = math.complex(3, 4);
 console.log(d.abs(), d.arg());      // radius = 5, phi = 0.9272952180016122
 
+// comparisions between two complex numbers (only considers real components)
+console.log('\ncomparisions between two complex numbers (only considers real components)');
```
Comment (josdejong) [inline examples/complex_numbers.js]: I think the comparison examples here are too comprehensive. Can you replace this with say two or the examples and a few lines of comments which explaining the behavior?

Comment (gulfaraz) [inline examples/complex_numbers.js]: https://github.com/josdejong/mathjs/pull/852/commits/4e324ee502148cdbf0768eed320601b37563dc6a should be concise enough

**Comment by gulfaraz** [comment]
Refer https://github.com/josdejong/mathjs/issues/837

**Comment by josdejong** [comment]
Wow, that's very fast, super! I think this is indeed the exact behavior we're looking for.

I'm not sure whether the abstraction with `Complex.significant` is the easiest one. Maybe it's easier to create a (static) compare function instead, like:

```js
  /**
   * Compare two complex numbers, `a` and `b`:
   *
   * - Returns 1 when the real part of `a` is larger than the real part of `b`
   * - Returns -1 when the real part of `a` is smaller than the real part of `b`
   * - Returns 1 when the real parts are equal 
   *   and the imaginary part of `a` is larger than the imaginary part of `b`
   * - Returns -1 when the real parts are equal
   *   and the imaginary part of `a` is smaller than the imaginary part of `b`
   * - Returns 0 when both real and imaginary parts are equal.
   *
   * @params {Complex} a
   * @params {Complex} b
   * @returns {number} Returns the comparison result: -1, 0, or 1
   */
  Complex.compare = function (a, b) {
    if (a.re > b.re) { return 1; }
    if (a.re < b.re) { return -1; }

    if (a.im > b.im) { return 1; }
    if (a.im < b.im) { return -1; }

    return 0;
  }
```

Which can be used like:

```js
  // larger
  'Complex, Complex': function (x, y) {
    return Complex.compare(x, y) > 0;
  },
```

what do you think?

**Comment by gulfaraz** [comment]
@josdejong Yeah I can do that, but wouldn't the name conflict with the other `compare` function ?

**Comment by josdejong** [comment]
>  wouldn't the name conflict with the other compare function ?

No it will be exactly the Complex implementation of the more generic `math.compare` function. You can even simplify your latest implementation:

```js
    // math.compare
    'Complex, Complex': function (x, y) {
      // use number implementation
      return type.Complex.compare(x, y);
    },
```

with:
```js
    // math.compare
    'Complex, Complex': type.Complex.compare,
```

:D 

Your latest update looks like we're there, though maybe you can cleanup the 
`// use number implementation` comments?

**Comment by josdejong** [comment]
Thanks @gulfaraz !

**Comment by gulfaraz** [comment]
Glad to help :) Thank you for the awesome library

**Comment by josdejong** [comment]
:+1: feel free to jump on other open issues if you see anything interesting (some are specifically marked with "help wanted")

**Pull request merged**